File: src/Explore.hs
module Explore
  ( runMission
  , runMissions
  , missionReport
  , Mission(..)
  , FailedMission(..)
  ) where

import           Control.Monad.State.Strict

import           Bot                        (Bot)
import qualified Bot
import           Commands
import           Headings                   (Heading)
import           Planet                     (Planet)
import qualified Planet

import           Data.Maybe                 (catMaybes)

newtype FailedMission =
  LostBot Bot
  deriving (Show, Eq)

type Exploration v = State Planet v

data Mission =
  Mission Int
          Int
          Heading
          String
  deriving (Show, Eq)

runMission :: Mission -> Exploration (Either FailedMission Bot)
runMission (Mission startX startY startHeading commandString) =
  let commands = catMaybes $ readCommand <$> commandString
      bot = Bot.create startX startY startHeading
  in do finalBot <- gets (\planet -> foldM (runCommand planet) bot commands)
        case finalBot of
          Right finalBot -> return $ Right finalBot
          l@(Left (LostBot bot)) -> do
            modify (\planet -> Planet.addMarker planet (Bot.position bot))
            return l

runCommand :: Planet -> Bot -> Command -> Either FailedMission Bot
runCommand planet bot command
  | Planet.checkCoords planet (Bot.position nextBot) = Right nextBot
  | Planet.checkMarker planet (Bot.position bot) = Right bot
  | otherwise = Left $ LostBot bot
  where
    nextBot = Bot.command bot command

runMissions :: [Mission] -> Exploration [Either FailedMission Bot]
runMissions = mapM runMission

missionReport :: Either FailedMission Bot -> String
missionReport (Right bot)          = show bot
missionReport (Left (LostBot bot)) = show bot ++ " LOST"

File: scala/src/main/scala/Explore.scala
package com.rumblesan.marsrover

import cats._
import cats.data._
import cats.instances.list._

import Mission.{MissionResult, FailedMissionResult, LostBot}


object Explore {

  type Exploration[Result] = State[Planet, Result]

  type MultipleExplorations[Result] = List[Exploration[Result]]

  def run(planet: Planet, missions: List[Mission]): List[MissionResult[Bot]] = {
    setup(missions).runA(planet).value
  }

  def setup(missions: List[Mission]): Exploration[List[MissionResult[Bot]]] = {
    Monad[Exploration].sequence[List, MissionResult[Bot]](
      missions.map((mission) => {
        for {
          missionResult <- State.inspect[Planet, MissionResult[Bot]](planet => Mission.run(mission, planet))
          result <- handleChanges(missionResult)
        } yield result
      })
    )
  }

  def handleChanges(result: MissionResult[Bot]): Exploration[MissionResult[Bot]] = {
    result match {
      case Left(LostBot(lostBot)) => {
        for {
          _ <- State.modify[Planet](planet => Planet.addMarker(planet, lostBot.xPos, lostBot.yPos))
        } yield result
      }
      case Right(bot) => State.pure(result)
    }
  }

}

